python/cudf/cudf/core/multiindex.py

```diff
@@ -103,20 +103,32 @@ def __init__(
         self._levels = [Series(level) for level in levels]
         self._validate_levels_and_codes(self._levels, self._codes)
 
-        source_data = DataFrame()
-        for i, name in enumerate(self._codes.columns):
-            codes = as_index(self._codes[name]._column)
-            if -1 in self._codes[name].values:
-                # Must account for null(s) in _source_data column
+        pandas_multindex = kwargs.get("pandas", None)
+
+        if pandas_multindex is None:
+            source_data = DataFrame()
+            for i, name in enumerate(self._codes.columns):
+                codes = as_index(self._codes[name]._column)
+                if -1 in self._codes[name].values:
+                    # Must account for null(s) in _source_data column
+                    level = DataFrame(
+                        {name: [None] + list(self._levels[i])},
+                        index=range(-1, len(self._levels[i])),
+                    )
+                else:
+                    level = DataFrame({name: self._levels[i]})
+
                 level = DataFrame(
-                    {name: [None] + list(self._levels[i])},
-                    index=range(-1, len(self._levels[i])),
+                    {"original_order": self.codes[name].index}, index=codes
+                ).join(level)
+                level = level.sort_values(by="original_order").reset_index(
+                    True
                 )
-            else:
-                level = DataFrame({name: self._levels[i]})
-            level = DataFrame(index=codes).join(level)
-            source_data[name] = level[name].reset_index(drop=True)
-
+                source_data[name] = level[name].reset_index(drop=True)
```

Comment:
Instead of doing a join + sort I think this can just be a `gather`. Where we can run a gather against `self._levels[i]` using `self.codes[name]` as the gather map.